### ISSUE TITLE
Training Process

### ISSUE THREAD

**Issue by kenkalang** [magronp/bsrnn#1]
Hi,

I am trying to train this model to separate guitar and piano, but the training process takes 3 hours per epoch with 12 num_repeats and 128 feature. The validation dataset is only 8 songs. Im using V100 32GB x1. Is there any way i could speed up the training process?

**Comment by magronp** [comment]
Hi,

The training time mostly depends on the size of your training dataset (not really on the validation set) and your available computing power.
In my case, using the large model (num_repeats=12 / feature_dim=128), two 45GB GPUs, and 20000 training samples (batch_size=2), one epoch is about 2h. I'm not sure what your training dataset is, but it seems like a comparable time.
I'm not aware of any way to speed things up, as the very nature of the model (LSTM) imposes a large memory usage and consequently a quite small batch size. I've tried replacing LSTMs with Conv1D, which is much faster, but the performance is quite poor.
Maybe start with a smaller model (e.g., num_repeats=8 / feature_dim=64), which allows you to increase the batch size. Let me know how it goes.

Best,
Paul

**Comment by kenkalang** [comment]
Thank you for the advice, Paul. I'll give it a try and provide an update once it's done.

**Comment by kenkalang** [comment]
Anyway I have a quick question: instead of reducing the feature_dim, would it be better to reduce the num_repeats to 8 while keeping the feature_dim at 128? In terms of training time and performance, which approach would likely be more effective?

Thanks again for your help!

**Comment by magronp** [comment]
In my experience, reducing `feature_dim` is what yields the most significant speed-up... but also the most significant drop in performance (compared to `num_repeat`).
Then, I think if you reduce `num_repeat`, you'll be able to roughly maintain performance, but it won't be much faster.
That's what I remember - but of course feel free to try it :)

**Comment by kenkalang** [comment]
I trained the guitar and piano models with the smaller model configuration you recommended, and the results are surprising. For the guitar model, I achieved 2.36 uSDR and 2.236 cSDR, compared to HT Demucs which had 2.46 uSDR and 2.161 cSDR on the same dataset. The piano model showed even better results, with 2.226 uSDR and 1.671 cSDR compared to HT Demucs’ 1.738 uSDR and 0.892 cSDR. I'm still working on training the other models (another 4 stems). And last thing, I noticed the mask estimation part is slightly different compared to the original paper. Is there a specific reason for this? Thank You

**Comment by magronp** [comment]
Looks like nice results compared to Demucs, although the values are quite low; maybe both models perform badly in this case?
I did a few modifications in the bsrnn model implementation (as mentionned [at the top of the script](https://github.com/magronp/bsrnn/blob/main/models/bsrnn.py)), but which part of the mask estimation are you referring to exactly?

**Comment by kenkalang** [comment]
In the paper's Mask Estimation module, they typically use MLP with GLU activation for the output layer. In your implementation, I see you've opted for a different strategy using Conv1d layers, with tanh activation applied within these layers. So what's the reason behind this?

Regarding performance, achieving such scores for guitar and piano is common. It's surprising that your implementation has achieved better results compared to Demucs. Keep up the good work!

**Comment by magronp** [comment]
You're right that the paper uses MLP for mask estimation, and my implementation is based on Conv1D!
I didn't spot that, since I simply the "official" BSRNN model implementation available [here](https://gitlab.aicrowd.com/Tomasyu/sdx-2023-music-demixing-track-starter-kit/-/blob/master/my_submission/bsrnn.py?ref_type=heads), which (according to the authors) was supposed to be the same one as in the paper... But obviously it's different.
I'll try to implement and test this layer norm + MLP module, and let you know how it goes.
Thanks a lot for pointing that out :)

EDIT: it seems like it aslo applies to the band-split module (FC in the paper vs. Conv1D it the authors'/my implementation). I'll check that as well.

**Comment by kenkalang** [comment]
It turns out that Conv1D and fully connected layers (FC) produce the same output, just using different methods. You can see it in this report: [FC Layer and Conv1d](https://wandb.ai/wandb_fc/pytorch-image-models/reports/Are-fully-connected-and-convolution-layers-equivalent-If-so-how---Vmlldzo4NDgwNjY#:~:text=To%20have%20the%20equivalence%20to,Linear%20layer%20operations%20are%20equivalent.)

**Comment by magronp** [comment]
Yes, just figured that out as well :) It's mostly a matter of initialization / speed.
Regarding the output activation (GLU): it's actually computed in the forward pass: `this_mask = this_output[:, 0] * torch.sigmoid(this_output[:, 1]`.
So it seems consistent with the paper so far.

**Comment by kenkalang** [comment]
Got it, thanks for the explanation.
